server/config.go

```diff
@@ -0,0 +1,185 @@
+package server
+
+import (
+	"encoding/json"
+	"errors"
+	"github.com/flashbots/go-boost-utils/types"
+	"os"
+)
+
+type rawConfiguration struct {
+	FeeRecipient          string `json:"fee_recipient"`
+	ValidatorRegistration struct {
+		BuilderRelays []string `json:"builder_relays"`
+		Enabled       bool     `json:"enabled"`
+		GasLimit      string   `json:"gas_limit"`
+	} `json:"validator_registration"`
+}
+
+type rawConfigurationFile struct {
+	BuilderRelaysGroups map[string][]string `json:"builder_relays_groups"`
+
+	ProposerConfig map[string]rawConfiguration `json:"proposer_config"`
+	DefaultConfig  rawConfiguration            `json:"default_config"`
+}
+
+// newRawConfigurationFile creates a temporary rawConfigurationFile used to build a
+// ProposerConfigurationStorage, by reading content from a JSON file.
+func newRawConfigurationFile(filename string) (*rawConfigurationFile, error) {
+	// Read JSON file content.
+	bytes, err := os.ReadFile(filename)
+	if err != nil {
+		return nil, err
+	}
+
+	// Tries to unmarshal content in JSON.
+	placeholder := &rawConfigurationFile{}
+	if err := json.Unmarshal(bytes, placeholder); err != nil {
+		return nil, err
+	}
+
+	return placeholder, nil
+}
+
+// ProposerConfig holds one proposer configuration.
+type ProposerConfig struct {
+	FeeRecipient types.Address
+	Enabled      bool
+	Relays       []RelayEntry
+	GasLimit     types.U256Str
+}
+
+// ProposerConfigurationStorage holds both the default configuration and the proposers ones.
+type ProposerConfigurationStorage struct {
+	proposerConfigurations map[types.PublicKey]*ProposerConfig
+	defaultConfiguration   *ProposerConfig
+}
+
+// NewProposerConfigurationStorage creates a new storage holding each proposer preferences using
+// the content extracted from a JSON file.
+func NewProposerConfigurationStorage(filename string) (*ProposerConfigurationStorage, error) {
+	// Tries to create the raw configuration extracted from the JSON file.
+	raw, err := newRawConfigurationFile(filename)
+	if err != nil {
+		return nil, err
+	}
+
+	// Initialize the storage and save default configuration.
+	pcs := &ProposerConfigurationStorage{
+		proposerConfigurations: map[types.PublicKey]*ProposerConfig{},
+	}
+
+	pcs.defaultConfiguration, err = newConfigurationStorage(&raw.DefaultConfig, raw.BuilderRelaysGroups)
+	if err != nil {
+		return nil, err
+	}
+
+	// For each proposer, save its own configuration.
+	for proposer, configuration := range raw.ProposerConfig {
+		address, err := types.HexToPubkey(proposer)
+		if err != nil {
+			return nil, err
+		}
+
+		configurationStorage, err := newConfigurationStorage(&configuration, raw.BuilderRelaysGroups)
+		if err != nil {
+			return nil, err
+		}
+
+		pcs.proposerConfigurations[address] = configurationStorage
+	}
+
+	return pcs, nil
```

Comment:
How can we dynamically update the configuration without restarting MEV-Boost to force the configuration file to be reread?

Reply:
Same question.
For example, https://github.com/spf13/viper provides out of the box dynamically updates config files without restarting goLang app.